cuda/btx_cudainterval_callbacks.cpp

```diff
@@ -0,0 +1,589 @@
+                                  BACKEND_CUDA, fn_name.c_str(), delta,
+                                  dev, dev, err, "");
+  } catch(const std::out_of_range& oor) {
+    THAPI_FATAL_HPT(hpt,
+      "device not found for hpt function name in profiling results callback");
+  }
+}
+
+
+// ===============================
+// Context management
+// ===============================
+void primary_ctx_retain_entry_callback(void *btx_handle, void *usr_data,
+                                       int64_t ts, const char *event_class_name,
+                                       const char *hostname, int64_t vpid,
+                                       uint64_t vtid, CUdevice dev) {
+  auto state = static_cast<data_t *>(usr_data);
+  state->context_manager.primary_ctx_retain_entry({hostname, vpid, vtid}, dev);
+}
+
+void primary_ctx_retain_exit_callback(void *btx_handle, void *usr_data,
+                                      int64_t ts, const char *event_class_name,
+                                      const char *hostname, int64_t vpid,
+                                      uint64_t vtid, CUresult cuResult,
+                                      CUcontext ctx) {
+  auto state = static_cast<data_t *>(usr_data);
+  state->context_manager.primary_ctx_retain_exit({hostname, vpid, vtid}, cuResult, ctx);
+}
+
+void primary_ctx_release_entry_callback(void *btx_handle, void *usr_data,
+                                        int64_t ts, const char *event_class_name,
+                                        const char *hostname, int64_t vpid,
+                                        uint64_t vtid, CUdevice dev) {
+  auto state = static_cast<data_t *>(usr_data);
+  state->context_manager.primary_ctx_release_entry({hostname, vpid, vtid}, dev);
+}
+
+void primary_ctx_release_exit_callback(void *btx_handle, void *usr_data,
+                                       int64_t ts, const char *event_class_name,
+                                       const char *hostname, int64_t vpid,
+                                       uint64_t vtid, CUresult cuResult) {
+  auto state = static_cast<data_t *>(usr_data);
+  state->context_manager.primary_ctx_release_exit({hostname, vpid, vtid}, cuResult);
+}
+
+void primary_ctx_reset_entry_callback(void *btx_handle, void *usr_data,
+                                      int64_t ts, const char *event_class_name,
+                                      const char *hostname, int64_t vpid,
+                                      uint64_t vtid, CUdevice dev) {
+  auto state = static_cast<data_t *>(usr_data);
+  state->context_manager.primary_ctx_reset_entry({hostname, vpid, vtid}, dev);
+}
+
+void primary_ctx_reset_exit_callback(void *btx_handle, void *usr_data,
+                                     int64_t ts, const char *event_class_name,
+                                     const char *hostname, int64_t vpid,
+                                     uint64_t vtid, CUresult cuResult) {
+  auto state = static_cast<data_t *>(usr_data);
```

Comment:
Should we handle the `cu_result` outside of the Context_manager? 
It will mimic what we do for other function. Not any strong preference

Reply:
I can see the logic of doing it outside context manager, but don't have a strong preference either, so inclined to leave it as is.